tests/unit/mixins/should-add-token-test.js

```diff
@@ -47,3 +46,59 @@ test('does throw an exception with bad input', function (assert) {
   assert.notOk(subject.hostsMatch(undefined, 'https://server.foo.bar.com:6443/FeatureServer/0'));
   assert.notOk(subject.hostsMatch(undefined, {}));
 });
+
+test('shouldAddToken returns false when token should not be added', function (assert) {
+  const url = 'https://services2.arcgis.com/zNjnZafDYCAJAbN0/arcgis/rest/services/Street_ROW_Trees/FeatureServer/0/query?returnCountOnly=true&f=json&where=1%3D1&inSR=4326&geometry=-118.132,34.16,-118.122,34.162&geometryType=esriGeometryEnvelope&spatialRel=esriSpatialRelIntersects';
+  const serverInfo = {
+    owningSystemUrl: 'http://www.arcgis.com',
+    owningTenant: 'zNjnZafDYCAJAbN0'
+  };
+  const portalInfo = {
+    portalHostname: 'qaext.arcgis.com',
+    id: '97KLIFOSt5CxbiRI',
+    authorizedCrossOriginDomains: []
+  };
+  assert.notOk(subject.shouldAddToken(url, serverInfo, portalInfo));
+});
+
+test('shouldAddToken returns true when portal ids match', function (assert) {
+  const url = 'https://servicesqa.arcgis.com/97KLIFOSt5CxbiRI/ArcGIS/rest/services/integration_features_10001/FeatureServer/0/query?returnCountOnly=true&f=json&where=1%3D1&inSR=4326&geometry=-95.107,41.389,-95.025,41.4&geometryType=esriGeometryEnvelope&spatialRel=esriSpatialRelIntersects';
+  const serverInfo = {
+    owningSystemUrl: 'http://www.arcgis.com',
+    owningTenant: '97KLIFOSt5CxbiRI'
+  };
+  const portalInfo = {
+    portalHostname: 'qaext.arcgis.com',
+    id: '97KLIFOSt5CxbiRI',
+    authorizedCrossOriginDomains: []
+  };
+  assert.ok(subject.shouldAddToken(url, serverInfo, portalInfo));
+});
+
+test('shouldAddToken returns true when hostnames match', function (assert) {
+  const url = 'https://servicesqa.arcgis.com/97KLIFOSt5CxbiRI/ArcGIS/rest/services/integration_features_10001/FeatureServer/0/query?returnCountOnly=true&f=json&where=1%3D1&inSR=4326&geometry=-95.107,41.389,-95.025,41.4&geometryType=esriGeometryEnvelope&spatialRel=esriSpatialRelIntersects';
+  const serverInfo = {
+    owningSystemUrl: 'http://www.arcgis.com',
+    owningTenant: 'zNjnZafDYCAJAbN0'
+  };
+  const portalInfo = {
+    portalHostname: 'www.arcgis.com',
+    id: '97KLIFOSt5CxbiRI',
+    authorizedCrossOriginDomains: []
+  };
+  assert.ok(subject.shouldAddToken(url, serverInfo, portalInfo));
```

Comment:
same thing, how is this getting around the authorizedCrossOrigin check?

Reply:
In this case it is because `serverInfo.owningSystemUrl ~~~ portalInfo.portalHostname`.

Reply:
I'm not sure that logic is correct. I think the whole domains need to match unless it's either an ArcGIS domain or authorized crossOrigin. Another way to look at this would be to automatically add the relevant arcgis domains to the crossOrigin list and only allow authorized cross origin or same origin

Reply:
That's what this checks - that the whole domains match. this is actually the part that I am pretty certain is correct.

Reply:
but the issue here is that the service url is neither part of the same domain as the portal, nor listed in authorizedCrossOrigin

Reply:
For reference, these two tests are based on actual use cases in opendata-ui:
`shouldAddToken returns false when token should not be added`
`shouldAddToken returns true when hostnames match`

Reply:
the end result is definitely correct. I'm just trying to understand how you get there without either a crossOrigin check or a whitelist for arcgis.com domains as I'm recommending

Reply:
It seems to me that if we whitelist arcgis.com domains, it would cause this test to fail: `shouldAddToken returns false when token should not be added`.

Reply:
how else are you going to allow services.arcgis.com utilties.arcgis.com landscape.arcgis.com, all of which don't appear in the crossOrigin list, but are still valid?

Reply:
I think you just need to skip the crossOrigin check if the url is *.arcgis.com, you can still match the owningSystem to make sure you are sending the correct token.